Transfers: add average transfer wait time metric #5012

davidcheng02 opened the pull request:
We keep track of the running average wait time for all completed transfers sent to update_transfer_state. This metric aims to measure the effectiveness of load distribution from different source node selection strategies. Lower average wait time generally signifies a more effective strategy. The average is based on a tumbling window of 6 hours.

<!-- Please read https://rucio.cern.ch/documentation/contributing before submitting a pull request -->

davidcheng02 commented on lib/rucio/core/request.py:
```diff
@@ -1390,6 +1396,20 @@ def observe(
+
+                        # avg_wait_time averages across a tumbling window, so reset avg every few hours
+                        hours_window = 6
+                        if self.avg_wait_time_start_timestamp < (now - datetime.timedelta(hours=hours_window)):
```
We decided to use tumbling window for calculating our running average. While sliding window may make more sense, it requires storing all the transfers in the past X hours, and additional computation is required to remove old transfers and adding in new transfers. Therefore, tumbling window is the most feasible.

davidcheng02 commented on lib/rucio/core/request.py:
```diff
@@ -1390,6 +1396,20 @@ def observe(
+
+                        self.avg_wait_time = ((self.avg_wait_time * self.transfers_done) + wait_time) / (self.transfers_done + 1)
+                        self.transfers_done += 1
+                        METRICS.gauge('avg_wait_time').set(self.avg_wait_time)
```
We do not add our gauge metric with any filters (such as by src id or activity), since this gauge will be a global value corresponding to the overall average wait time for an entire topology. We imagine that on prometheus, this gauge will show up as a line graph or time plot, where the x-axis is the timestamp when the gauge metric was pulled, and y-axis is the avg_wait_time.

rcarpa commented:
As mentioned off-line, you should use a prometheus Histogram (It's badly named "Timing"/"timer" in our code for historical reasons linked to usage of statsd in the past). The usage of histograms is a bit counter-intuitive and tricky. Mostly because you have to correctly set the "buckets", so you'll have to learn how to use it. The advantage is that we don't need to compute a moving average in the code. We can record raw values. If the buckets are correctly chosen, we'll be able to use prometheus integrated query engine to answer questions like "how many requests which ended in the last 24 hours took more than a hour to complete" or "what is the duration of transfers if we ignore the highest 5 percentile" .. etc.

In this case buckets will have to be set in a way which cover both the shortest transfers and the longest ones. For example: 
`10s 30s 1m 5m 10m 20m 40m 1h30m 3h 6h 12h 1d 2d 3d 4d 5d 6d 7d 10d 14d 1m infinity`. You have quite sparse granularity at higher ranges, but it should be "good enough"

davidcheng02 committed:
Transfers: add average transfer wait time metric #5012
We keep track of the wait time and transfer time for each completed
transfer sent to update_transfer_state. This metric aims to measure the
effectiveness of load distribution from different source node selection
strategies. Lower average wait time and transfer time generally
signifies a more effective strategy.

rcarpa commented:
The test failure at https://github.com/rucio/rucio/blob/436ecc9005934e1e4332cd2718d160bdbf5105d2/tests/test_conveyor.py#L287 may actually be a sign that the new code raises an exception. This has to be double-checked .
